src/main/scala/loamstream/compiler/LoamCompiler.scala

```diff
@@ -287,15 +295,45 @@ final class LoamCompiler(
     f(new compiler.Run)
   }
 
-  private def depositProjectContextAndThen[A](project: LoamProject)(f: DepositBox.Receipt => A): A = {
-    val projectContextReceipt = LoamProjectContext.depositBox.deposit(LoamProjectContext.empty(project.config))
-
-    try { f(projectContextReceipt) }
-    finally {
-      LoamProjectContext.depositBox.remove(projectContextReceipt)
+  /**
+   * Make a new LoamProjectContext, set the value of LoamFile.ContextHolder.projectContext with it, and run
+   * f.  Since LoamFile.ContextHolder.projectContext uses a ThreadLocal to store the passed value, do the above
+   * in a new Thread to ensure that (effectively) each Loam compilation/evaluation gets its own LoamProjectContext,
+   * as before.  When f completes, clear out the LoamFile.ContextHolder.projectContext ThreadLocal.
+   */
+  private def setProjectContextAndThen[A](project: LoamProject)(f: LoamProjectContext => A): A = {
+    doInNewThread {
+      try {
+        val projectContext = LoamProjectContext.empty(project.config)
+  
+        LoamFile.ContextHolder.projectContext = projectContext
+  
+        f(projectContext)
+      } finally {
+        LoamFile.ContextHolder.clear()
+      }
     }
   }
 
+  /**
+   * Synchronously run the passed code block, but do it in a new Thread.
+   */
+  private def doInNewThread[A](block: => A): A = {
```

Comment:
This function feels wonky. Why would I do something in a new thread and then block waiting on it instead of just doing that thing in this thread. It has the same effect and doesn't require context switching or OS handles. 

There's likely a case this is useful/valuable I'm unaware of. Might be good adding to the comment where it's valuable to do.

Reply:
It is an odd thing to do, no question.  The comment for `setProjectContextAndThen` explains in more detail: 
```
Make a new LoamProjectContext, set the value of LoamFile.ContextHolder.projectContext with it, and run
   * f.  Since LoamFile.ContextHolder.projectContext uses a ThreadLocal to store the passed value, do the above
   * in a new Thread to ensure that (effectively) each Loam compilation/evaluation gets its own LoamProjectContext,
   * as before.  When f completes, clear out the LoamFile.ContextHolder.projectContext ThreadLocal.
```

This is necessary to allow making "global" state for a whole compilation-evaluation run available to a bunch of `object`s.  Using a `ThreadLocal` in a known place achieves "global" state, while the `ThreadLocal` allows the state to actually be per compilation-evaluation run.  For this to work, each compilation-evaluation run must happen in a new thread, hence this method.  

This is not my favorite thing, and am open to suggestions.  I appreciate the pointer to `DynamicVariable`, which I did not know about before.